### PR TITLE
hostapd: Disable 802.11b data rates by default

### PR DESCRIPTION
**Pull request by nicklowe**
Set legacy_rates to 0 by default to disable 802.11b data rates by default.

The time has long come where 802.11b DSSS/CCK data rates should be disabled
by default in OpenWRT. Users in need of 802.11b client support can reasonably
enable these where they are needed.

The balance of equities has significantly, and for a long time, tipped
such that dropping backwards compatibility by default with 802.11b
devices is appropriate, proportionate and justified. By doing so,
management and control traffic is moved by default to a 20
MHz wide 6 Mb/s OFDM data rate instead of a 22 MHz wide 1 Mb/s DSSS data
rate. This is significantly more airtime efficient.

Signed-off-by: Nick Lowe <nick.lowe@gmail.com>

### CODE REVIEW

**Review by dangowrt — APPROVED** [review]

**Comment by adschm** [comment]
Please bump PKG_RELEASE.

**Comment by fblaese** [comment]
I think this is a good idea. I haven't seen any devices only capable of 802.11b for a really long time now, and the 22 MHz wide DSSS data rates waste precious bandwidth, which could be used for more tightly packed channels (1, 6, 11 vs 1, 5, 9, 13).

**Comment by jow-** [comment]
Please also do a corresponding PR aligning the default in LuCI.

**Comment by nicklowe** [comment]
The companion PR for LuCI can be found here: https://github.com/openwrt/luci/pull/4637

**Comment by Borromini** [comment]
While I fully understand the rationale, disabling them has ugly consequences. What you will see is people complaining they cannot connect to their OpenWrt router.

I set up friends of mine with an 802.11ac router, disabled 802.11b, and thought they'd be just fine with all their hardware being either ac or (b)gn. But a few weeks after they told me their bgn laptops could not connect anymore. Took me a while to see the link between disabling 802.11b and that issue. The signal is still more or less okay, so I'm not sure what's at play, but enabling the legacy rates again made the issue go away. Not sure if competition for airtime with neighbouring networks is a factor here either.

Just my two cents :smiley:

**Comment by adschm** [comment]
> I set up friends of mine with an 802.11ac router, disabled 802.11b, and thought they'd be just fine with all their hardware being either ac or (b)gn. But a few weeks after they told me their bgn laptops could not connect anymore. Took me a while to see the link between disabling 802.11b and that issue. The signal is still more or less okay, so I'm not sure what's at play, but enabling the legacy rates again made the issue go away. Not sure if competition for airtime with neighbouring networks is a factor here either.

I went through the same misery myself and at my parents home recently. First, I ended up enabling 11b again to have all devices working. But still, this is a rather unsatisfying state, and is not an option if you want to do proper 1/5/9/13. So, eventually, I bought one (cheap) dual band adapter and found one draft-n (!) adapter in box, both replacing the two devices actually making problems. Now, everything is fine on both households.

So, while I understand the concern of people not having all of their devices working, I think it's time now to use this as a chance to phase out all those ancient, broken (!) devices that are making problems here. If you buy something new, you might get an additional radio for almost the same price as a collateral benefit ...

**Comment by blocktrron** [comment]
While i have it personally enabled (and i would advice anyone to do so), i have a strong opinion on leaving the DSSS rates enabled. There's still to much IoT stuff in productions which either breaks on missing DSSS rates for the Basic rates or expect beacons to be transmitted at the lowest possible rate.
 
Just some of which I've personally fought with this year (still in production):

 > Panasonic CS-TZ25WKEW
 > HP Laser 107W
 > Silex UART WiFi modules 

So this is a NAK from my side.

**Comment by nicklowe** [comment]
For some context to this discussion, the question of disabling DSSS rates was put to the large deployment-supporting wireless community on Twitter a few years back and the feedback was pretty unanimous: https://twitter.com/wirednot/status/662582711953944576?s=20

Many enterprise/business environments are already configured this way and have been successfully for years. Clearly here  though we would also be bringing more in to scope SOHO and home environments and the different type of devices they have, and therefore there is some risk.

However, I suggest to keep in mind that 802.11b support would not be removed here, just disabled by default and the option to enable remains.

My suggestion would be to make this change and take mitigatory steps such as clearly mentioning this in the release notes and also in documentation so that there is good upfront visibility and awareness.

For information, Cisco put a presentation forward to the IEEE 6 years ago seeking much the same goal as we would be here:
https://mentor.ieee.org/802.11/dcn/14/11-14-0099-00-000m-renewing-2-4ghz-band.pptx

If both SOHO and home-targeted manufacturers and the open-source space do not take the first step and change this, taking some risk in doing so, how do things start to improve in the 2.4 GHz junk band, particularly in congested environments?

I suggest that the way that OpenWRT ships today out of the box is not good for congested everyday environments where there are multiple BSSes sharing a channel. Leaving these enabled is therefore the wrong balance of equities in my opinion. The overheads from the beacons alone in such environments is explained well here: http://revolutionwifi.blogspot.com/p/ssid-overhead-calculator.html

**Comment by nicklowe** [comment]
Taking in to account David Bauer's legitimate concerns, I have updated the proposed LuCI patch this morning for discoverability:

1) Moving the option from the Advanced Settings to the General Setup tab.
2) Adding a description explaining potential compatibility implications. This states:

**Poorly behaving 802.11g or 802.11n devices may require legacy 802.11b rates to interoperate. Airtime efficiency may be significantly reduced where these are used. It is recommended to disable this option where possible.**

**Comment by adschm** [comment]
> I suggest that the way that OpenWRT ships today out of the box is not good for congested everyday environments where there are multiple BSSes sharing a channel. Leaving these enabled is therefore the wrong balance of equities in my opinion. 

I think this is a strong point. And I believe the people dealing with and using a lot of wireless hardware in their homes (including mentioned poorly implemented/old devices) are more likely to enable 11b rates again than the people just installing OpenWrt are likely to disable 11b for fairness only.

**Comment by aparcar** [comment]
I'm in favor of this. It should be added to the 20.x release information. This is about the default behaviour, not about removing a driver/implementation.

**Review by aparcar — APPROVED** [review]

**Comment by neheb** [comment]
A middle ground would be to disable by default for 5ghz networks.

Anyway, I have it disabled as well. I have no issues.

**Comment by aparcar** [comment]
Merged, thank you.

**Comment by SvenRoederer** [comment]
Happy to see this, our Firmware has been carrying this around for 2,5 years now (https://github.com/freifunk-berlin/firmware/commit/3a59c453c8d4dadd40d26e88f642a07acb9c074c).
Time to drop our patch for this.

**Comment by aparcar** [comment]
@SvenRoederer I see you have a bunch of other patches there, please feel free to upstream whatever seems fitting.

**Comment by adschm** [comment]
This has been explicitly NAKed by a committer, so I don't think we can just close the discussion here at this point.

**Comment by aparcar** [comment]
@blocktrron sorry I missed your explicit NAK under the device list and thought the follow up comments of @nicklowe and @adschm resolved the issue.

**Comment by SvenRoederer** [comment]
As mentioned before I'm a fan of disabling b-rates by default, too. In our community we have it off by default since our 1st LEDE based release. It's running on more >300 nodes of our Freifunk-community and I'm not aware of any user complains.

**Comment by dangowrt** [comment]
> A middle ground would be to disable by default for 5ghz networks.

Has there ever been CCK encoding on 5GHz?

**Comment by neheb** [comment]
I doubt it. 802.11a == 802.11g on 5GHz.

**Comment by nicklowe** [comment]
> Has there ever been CCK encoding on 5GHz?

> I doubt it. 802.11a == 802.11g on 5GHz.

Yes, that is exactly the case. There is a good overview here of the generations:
https://www.duckware.com/tech/wifi-in-the-us.html#legacy

DSSS/CCK data rates are just by 802.11 spec used for the 2.4 GHz band.

**Comment by adschm** [comment]
> A middle ground would be to disable by default for 5ghz networks.

I think that would just make everything more complicated.

**Comment by blocktrron** [comment]
I understand and support the deactivation of DSSS rates on larger deployments (Did it myself at work and at home).

However all of this people opted in to. I don't see OpenWrt targeting these deployments ootb. When we talk about the defaults we ship, we should strive for maximum compatibility, at least this is how i see it.

Yes, these client devices do not behave according to the IEEE spec. On the other hand, the WFA explicitly states WiFi certified products have to support all prior standards OOTB for compatibility. I'm aware we do not have to adhere to their specifications, but in the end this is what most vendors do and expect to work with (sadly).

Problems only get worse as soon as you introduce the breakage created by fullmac drivers which decide to ignore the management rate you've chosen to use and begin sending out beacons on a 1mbit/s rate which exclude this rate in their Basic rate set.

See my post on the ath10k list from earlier this year:

```
To: ath10k@lists.infradead.org
From: David Bauer <mail@david-bauer.net>
Subject: Management rate-control on IPQ4019
Message-ID: <20543d04-7847-8a4f-7882-78c2f42c98c0@david-bauer.net>
Date: Tue, 18 Feb 2020 20:12:59 +0100
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101
 Thunderbird/68.5.0
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Language: en-US
Content-Transfer-Encoding: 8bit

Hello,

while playing around with the 2.4GHz WiFi part of the IPQ4019, i was 
expecting being able to set the rate at which IPQ4019 transmits it's beacon
frames.

Using OpenWrt, setting "legacy_rates=0" on the radio leads to only advertising
802.11g speeds, however the beacons are still sent out at 1 Mbit/s. Using a QCA9984,
the beacons are correctly sent out at the lowest 802.11g rate (6 Mbit/s). So i assume
this is either a bug in the ath10k firmware or a hardware-shortcoming. Has anyone else
experienced this bug and is it likely we'll see it fixed in a later firmware release?

Hardware: IPQ4029 (Aruba AP-303)
Firmware  Version: 10.4-3.6-00140 / 10.4-3.5.3-00078

Best wishes
David
```

Don't get me wrong, I'm all in for optimizations done here (and I liked the cell_density option), but this is not the state in which i want to ship OpenWrt as an OOTB package, especially not in form of this rollercoaster rushed decision making process on a single weekend.

**Comment by nicklowe** [comment]
> When we talk about the defaults we ship, we should strive for maximum compatibility, at least this is how i see it.

I suggest that leaving 802.11b data rates enabled by default would seek compatibility ‘by default’ with what are very edge case, unusual and infrequently encountered clients. The cost of that is a significant and demonstrable detriment to the wider wireless ecosystem.

I kindly suggest this is not the right conclusion and is the wrong balance of equities. It seems to me to be an approach that seeks maximum compatibility no matter what the costs, I cannot agree with that. Here, we have significant, well understood, real world costs that I suggest significantly outweigh the compatibility concern.

It is great that the IPQ4029 bug has been identified! Has it been addressed and what is its current status?

Unarguably, and I do not dispute that there can be some inconveniences caused as latent bugs then bubble up and come to attention as a result of making this change, either on the AP or client/STA side.

Isn't that a good thing though as they can then be worked on and addressed?

Otherwise, I suggest we easily mask the issues and we let ourselves be trapped and held hostage by a 1999 protocol in 2020.

**Comment by blocktrron** [comment]
> It is great that the IPQ4029 bug has been identified! Has it been addressed and what is its current status?

As ath10k seems to be service-only by now, i don't think there is much hope left. For ath10k-ct there seems to be a sysfs knob, not sure if this works with the non-ct firmware (It's only described in context of wave 1). There was one constellation which resulted in a correct rate selection with the -ct firmware and non-ct driver. But again, this was at the beginning of the year.

Even if this issue with ath10k is fixed for all chipsets, mwlwifi openly admits it has the same shortcoming (https://github.com/kaloz/mwlwifi/blob/382700ce5744fe80271c57a89c6589e767d91620/mac80211.c#L409)

And I have to correct myself, this behavior is NOT in line with the IEEE specification which explicitly states the following:

Screenshot from 2020-12-08 17-14-37

While now we have this:

Screenshot from 2020-12-08 17-47-01

Maybe I'm interpreting the specification wrong at this point, if so please correct me here.

If we can tweak this situation for devices we still support while not violating the spec I'm fine with this PR. But in the state this is on various devices we still support, I really dislike shipping this per default.

**Comment by adschm** [comment]
@aparcar As you see in the discussion, the NAK is still kept up (and justified ...), so IMO the patch should be reverted in master (at least for now).

**Comment by nicklowe** [comment]
Hi,

>Maybe I'm interpreting the specification wrong at this point, if so please correct me here.

The 802.11 spec is referring to the MCS rates that are used with 11n, 11ac and 11ax.

HT = 802.11n
VHT = 802.11ac
HE = 802.11ax

It is therefore invalid to use rates from those 802.11 amendments for data and management frames.

OFDM rates are not prohibited here.

Best,

Nick

**Comment by nicklowe** [comment]
> @aparcar As you see in the discussion, the NAK is still kept up (and justified ...), so IMO the patch should be reverted in master (at least for now).

I respectfully disagree that it is justified. There is no functional, service affecting regression if a device actually sends its beacon frames at the 1 Mb/s rate. We just keep the airtime inefficiency on such devices until, if-and-when the bug can be fixed so that the desired rate can be used.

This is perfectly valid per-802.11 spec as explained.

Best,

Nick

**Comment by nicklowe** [comment]
Hi David,

I suggest for the issues highlighted with ath10k and IPQ4029 and mlwifi, we consider:

**What is the impact from this defaults change? Does anything functionally regress?**

As far as I can am aware and can see, nothing bad happens at all in that regard, there is no functional regression.

This then therefore becomes a separate and independent question of:

Can we fix those drivers and firmwares so that they start to work with the legacy_rates, cell_density, basic_rate and supported_rates settings properly?

Best,

Nick

**Comment by blocktrron** [comment]
> The 802.11 spec is referring to the MCS rates that are used with 11n, 11ac and 11ax.
> It is therefore invalid to use rates from those 802.11 amendments for management and control frames.
> OFDM rates from 802.11a and 802.11g are not prohibited here.

I'm aware of this, however the 2016 specification in addition explicitly states the rate the frame is transmitted with has to be contained in the basic rate set (which it is clearly not).

The 2007 specification (predating 802.11n / STBC) to which the 2016 specification is backwards compatible says

Screenshot from 2020-12-08 22-40-03

Chapter 9.12 lists all possible frame exchange sequences including sending the beacon frame

Screenshot from 2020-12-08 22-41-24

So the beacon frame should be transmitted with a rate contained in the basic rate set, shouldn't it?

To sum it up (Unless my understanding of the specification is completely wrong): 

Yes, most drivers will end up in a state which complies with this, but the fact is we have SOME current ones which do not.

Yes, there are enough STAs out there which work w/o issues with this non-standard result.

What I'd like is the decision process be based on the question "Do we want the benefit of reduced Airtime at the cost of potentially breaking a few (<5% I'd guess) installation with the default configuration". If two other maintainers outvote me on this, I'd happily even merge this personally after a reasonable time frame (which 3 days without any approval after by NAK obviously isn't). I don't think a instant revert is not necessary, as long as two "decision makers" outvote me here, given the concerns I've voiced. Otherwise, I'd revert this in the coming weeks.

And to be clear, with "vote" i don't expect this to be dragged out in some "official" procedure on openwrt-adm.

**As a closing thought**

As already explained multiple times, from my personal POV, this change is resonable and fine. However, from my OpenWrt maintainer POV, potentially decreasing interoperability because we do not adhere to the standard is a big nono.

But, as long as others who have to make the decision from the same POV outweigh my concerns, I'm perfectly fine talking the path the majority agreed on (We are not killing kittens here, do we?).

**Comment by nicklowe** [comment]
Yes, to be clear, if-and-where beacon and probe response frames for a BSS do have the 6 Mb/s OFDM rate as the minimum basic rate and yet these are actually transmitted at 1 Mb/s DSSS rate due to an underlying bug or limitation in driver/firmware, this would not strictly comply with 802.11 spec.

Are you perchance aware of any clients, chipsets or drivers that are sensitive to this specifically?

The reason I ask is that none should be as frames are not filtered out based on the basic and supported rate values for a BSS at a late stage after demodulation based on the rate it was received at. This is not a thing as far as I am aware. As long as the radio is capable of handling the rate and hearing it, it will.

**Comment by mmortal03** [comment]
Sort of tangential, but within the last month or so, a Roku software update was pushed out that broke their devices' detection of WiFi signals from routers... when this setting is *enabled*. Who knows why, but unchecking it works around the issue and allows those Roku devices to detect the network SSIDs again. Interestingly, it's not just affecting people on OpenWrt, it just so happens that OpenWrt has this setting that we can uncheck to work around the problem. I have no idea if Roku bases any of their WiFi functionality on open source software, but it'd be interesting to know what specifically changed in their code that broke it.

**Comment by s-2** [comment]
> Yes, to be clear, if-and-where beacon and probe response frames for a BSS do have the 6 Mb/s OFDM rate as the minimum basic rate and yet these are actually transmitted at 1 Mb/s DSSS rate due to an underlying bug or limitation in driver/firmware, this would not strictly comply with 802.11 spec.

So, technically, rather than discussing the desired default behaviour, the setting should not be available to the user in the first place, whenever it is known that the respective driver does not (yet) correctly implement disabling the lower rates, thus violating the specs?

I wonder if IPQ4018 is affected here as well (we have lots of FB4040 at Freifunk Hannover), will probably have to check that...

**Comment by nicklowe** [comment]
> So, technically, rather than discussing the desired default behaviour, the setting should not be available to the user in the first place, whenever it is known that the respective driver does not (yet) correctly implement disabling the lower rates, thus violating the specs?

This certainly could be implemented this but I suggest it is not at all in any way worth the effort and the additional code bloat as client devices are just not sensitive to this. Filtering/gating frames out on RX after demodulation is just not a thing that is ever done by anybody. For example, to receive beacons and probe response frames in the first place that contain the basic and supported rates, there is first a need to demodulate the frame.

**Comment by ynezz** [comment]
Seems like this is tough one :) BTW I'm in favor of accepting it as well, the pros are more compelling.

@nicklowe Could you please bring this to the broader audience provided by the openwrt-devel mailing list? Sending this as a patch is good enough.

**Comment by ynezz** [comment]
I've brought this to the attention and included it as a topic for next developer meeting.

**Comment by nicklowe** [comment]
>@nicklowe Could you please bring this to the broader audience provided by the openwrt-devel mailing list? Sending this as a patch is good enough.

Sure, no problem. Will do so over the weekend.

>I've brought this to the attention and included it as a topic for next developer meeting.

Thanks!

I did previously confirm the actual transmitted rate with a Netgear X4S (R7800), ipq8065 using out-of the-box ath10k-ct and confirmed no issue with mismatching rates there:

Screenshot 2020-12-10 at 13 53 27

There is definitely now a need to seek out and identify all targets for which we do not see the correct wlan_radio.data_rate value in a monitor mode PCAP in Wireshark and to then endeavour to get this corrected.

I suggest, however, this can reasonably be a separate and independent effort, and therefore be non-blocking to this PR, as no clients are known to be sensitive to this, and it very hard to see how any could be sensitive.

Instead, affected targets would just remain with the airtime inefficient behaviour if-and-until fixed.

**Comment by richb-hanover** [comment]
Sorry for the late thought: I know the Make-Wifi-Fast project has patches for various Wi-Fi chips that implement "Airtime Fairness". This tweaks the entire Wi-Fi stack to give each station a fair amount of airtime (not an equal number of bits). So slow stations get a few bits during their time slice while faster stations send more bits during theirs. Toke Høiland-Jørgensen wrote his PhD thesis on this and used OpenWrt as a test bed. See [Ending the Anomaly: Achieving low latency and airtime fairness](https://www.cs.kau.se/tohojo/airtime-fairness/)

My question: how does/could this affect the discussion of turning off 80211b? Thanks.

**Comment by nicklowe** [comment]
>My question: how does/could this affect the discussion of turning off 80211b? Thanks.

Airtime fairness does not have a relation to this discussion, which instead is about efficient use of airtime.

**Comment by aparcar** [comment]
Based on the OpenWrt dev meeting we agreed on keeping this change. Again, I'm sorry for missed the NAK of @blocktrron and will read comments more carefully in the future.

**Pull request closed without merging**
